hera_cal/apply_cal.py

```diff
@@ -360,22 +405,51 @@ def apply_cal(data_infilename, data_outfilename, new_calibration, old_calibratio
                           data=data, flags=data_flags, add_to_history=add_to_history, clobber=clobber, **kwargs)
         else:
             all_red_antpairs = [[bl[:2] for bl in grp] for grp in all_reds if grp[-1][-1] == hd.pols[0]]
-            hd.update(data=data, flags=data_flags, **kwargs)
+            hd.update(data=data, flags=data_flags, nsamples=data_nsamples, **kwargs)
             # by default, weight by nsamples (but not flags). This prevents spectral structure from being introduced
             # and also allows us to compute redundant averaged vis in flagged channels (in case flags are spurious).
             if no_red_weights:
                 redundant_weights = copy.deepcopy(data_nsamples)
                 for bl in data_flags:
                     if np.all(data_flags[bl]):
                         redundant_weights[bl][:] = 0.
-
-            utils.red_average(hd, reds=all_red_antpairs, inplace=True, wgts=redundant_weights,
-                              propagate_flags=True)
-            if filetype_out == 'uvh5':
-                # overwrite original outfile with
-                hd.write_uvh5(data_outfilename, clobber=clobber)
-            else:
-                raise NotImplementedError("redundant averaging only supported for uvh5 outputs.")
+            for red_chunk in range(redundant_groups):
+                red_antpairs = []
+                reds_data_bls = []
+                for grp in reds_data:
+                    # trim group to only include baselines with redundant weights not equal to zero.
+                    grp0 = grp[0]
+                    if dont_red_average_flagged_data:
+                        grp = [ap for ap in grp if np.any(np.asarray([~data_flags[ap + (pol,)] for pol in data_flags.pols()]))]
+                    # only include groups with more elements then redundant groups!
+                    if len(grp) >= redundant_groups:
```

Comment:
Since these bls will be ordered, you may end up with groups that are on opposite sides of the array. Perhaps it's more sensible to just use something like `grp[red_chunk:: redundant_groups]`?

Reply:
I've updated the code to more evenly distribute baselines in these groups.